# Explore here

In [2]:
import pandas as pd 

# carga de datos 
Url =  "https://raw.githubusercontent.com/4GeeksAcademy/naive-bayes-project-tutorial/main/playstore_reviews.csv"
df = pd.read_csv(Url)

# eliminar columnas irrelevantes 
df.drop("package_name", axis=1, inplace=True)

# Limpieza
df["review"] = df["review"].str.strip().str.lower()

print(df.head())

                                              review  polarity
0  privacy at least put some option appear offlin...         0
1  messenger issues ever since the last update, i...         0
2  profile any time my wife or anybody has more t...         0
3  the new features suck for those of us who don'...         0
4  forced reload on uploading pic on replying com...         0


La previsualización de las primeras 5 filas valida que el dataset está limpio y listo para el proceso de Split.

In [ ]:
from sklearn.model_selection import train_test_split

# variables
X = df["review"]
y = df["polarity"]

# División 80/20
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Validación de dimensiones
print(f"Entrenamiento: {X_train.shape[0]} registros")
print(f"Prueba: {X_test.shape[0]} registros")

Entrenamiento: 712 registros
Prueba: 179 registros


Se aplicó un split de 80% para entrenamiento y 20% para evaluación, asegurando una base de datos suficiente para el aprendizaje del modelo.

In [ ]:
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.naive_bayes import MultinomialNB

# Preparación del vectorizador
vec_model = CountVectorizer(stop_words="english")

# Transformación de sets
X_train_vec = vec_model.fit_transform(X_train).toarray()
X_test_vec = vec_model.transform(X_test).toarray()

# Entrenamiento
model = MultinomialNB()
model.fit(X_train_vec, y_train)

# Salida 
print(f"Precisión en entrenamiento: {model.score(X_train_vec, y_train):.2%}")

Precisión en entrenamiento: 96.07%


El 96.07% de precisión confirma que el algoritmo MultinomialNB ha identificado con éxito los patrones de frecuencia de palabras asociados a cada polaridad (positiva/negativa).

In [5]:
from sklearn.metrics import classification_report, accuracy_score

# Realizar predicciones 
y_pred = model.predict(X_test_vec)

# Calcular precisión final
total_acc = accuracy_score(y_test, y_pred)

# reporte detallado
print(f"Precisión Final (Test): {total_acc:.2%}")
print("\nReporte de Clasificación:")
print(classification_report(y_test, y_pred))

Precisión Final (Test): 81.56%

Reporte de Clasificación:
              precision    recall  f1-score   support

           0       0.84      0.90      0.87       126
           1       0.73      0.60      0.66        53

    accuracy                           0.82       179
   macro avg       0.79      0.75      0.77       179
weighted avg       0.81      0.82      0.81       179



 Del 81.56%. Hay sobreajuste, ya que en entrenamiento dio 96.07%, El modelo es bueno detectando lo negativo (F1: 0.87), pero es mediocre con lo positivo (F1: 0.66). En conclusion el dataset está bien balanceado, tiene muchas más reseñas negativas que positivas, y el modelo aprendió con sesgo.

In [6]:
import joblib

# 1. Guardar modelo y vectorizador
joblib.dump(model, "nb_model.pkl")
joblib.dump(vec_model, "vectorizer.pkl")

print("Proyecto listo para entrega: Archivos .pkl generados.")


Proyecto listo para entrega: Archivos .pkl generados.


Se generaron los archivos nb_model.pkl y vectorizer.pkl, permitiendo ejecutar el modelo en producción sin reentrenar. El proyecto queda cerrado con una precisión del 81.56% y el empaquetado de las dependencias de preprocesamiento.